src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp

```diff
@@ -68,36 +68,46 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
   };
 
 private:
-  struct Link : public ArenaObj {
-    Link* next;
+  struct Link;
+  using Allocator = HomogenousObjectArray<Link, mtNMT>;
+  using LinkPtr = typename Allocator::I;
+  LinkPtr nil() { return Allocator::nil; }
+
+  Allocator _allocator;
```

Comment:
"Allocator" is misleading. An allocator is typically a thing that provides alloc() and free() for you, but does not maintain the storage for you, thats up to you. 

Proposal: Allocator -> TableEntryStorage or EntryStorage. _allocator = _entry_storage.